### PR TITLE
PSS events feed and websockets

### PR DESCRIPTION
**Pull request by nolash**
Adds pub/sub of incoming Pss messages using event.Feed, and a websockets interface for pss send and notifications through persistent connection using rpc.notifier.

Websocket RPC are started automatically on the swarm executable if the --pss flag is present. If more instances are to be run, alternate ws ports can be specified with the --pssport flag.

Current implementation also includes a shallow builtin ping/pong handler with topic "pss" and version 1, to simplify POCs and testing against pss. This will eventually be removed, so no code should be made dependent on it.

### CODE REVIEW

File: swarm/network/pss.go [hunk 1]
```diff
@@ -371,7 +403,9 @@ func (prw PssReadWriter) WriteMsg(msg p2p.Msg) error {
 
 	pmsg, _ := makeMsg(msg.Code, ifc)
 
-	return prw.Pss.Send(to, *prw.topic, pmsg)
+	prw.Pss.Send(to, *prw.topic, pmsg)
```
Comment (zelig) [inline swarm/network/pss.go]: why this change?

File: swarm/network/pss.go [hunk 2]
```diff
@@ -436,13 +470,24 @@ func (self *PssProtocol) handle(msg []byte, p *p2p.Peer, senderAddr []byte) erro
...
-func (ps *Pss) isSelfRecipient(msg *PssMsg) bool {
-	if bytes.Equal(msg.GetRecipient(), ps.Overlay.GetAddr().OverlayAddr()) {
+func (self *Pss) IsSelfRecipient(msg *PssMsg) bool {
+	if bytes.Equal(msg.GetRecipient(), self.Overlay.GetAddr().OverlayAddr()) {
 		return true
 	}
 	return false
 }
 
+func (self *Pss) GetPingHandler() func([]byte, *p2p.Peer, []byte) error {
```
Comment (zelig) [inline swarm/network/pss.go:478]: can we please move this to the test?

File: swarm/network/pssapi.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		
+	go func(topic PssTopic) {
+		for {
+			select {
+				case msg := <-ch:
+					if err := notifier.Notify(sub.ID, msg); err != nil {
+						log.Warn(fmt.Sprintf("notification on pss sub topic %v rpc (sub %v) msg %v failed!", topic, sub.ID, msg))
+					}
+				case err := <-psssub.Err():
+					log.Warn(fmt.Sprintf("caught subscription error in pss sub topic: %v", topic, err))
```
Comment (zelig) [inline swarm/network/pssapi.go]: why not return the error? lets return the error :)

File: swarm/network/pssapi.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+					return
+				case <- notifier.Closed():
+					log.Warn(fmt.Sprintf("rpc sub notifier closed"))
+					psssub.Unsubscribe()
+					return
+				case err := <- sub.Err():
```
Comment (zelig) [inline swarm/network/pssapi.go]: let's return the error

File: swarm/swarm.go [hunk 5]
```diff
@@ -199,20 +199,24 @@ func (self *Swarm) Start(net p2p.Server) error {
...
 	)
 
-	log.Info(fmt.Sprintf("Swarm network started on bzz address: %v", self.hive.Addr()))
+	log.Info(fmt.Sprintf("Swarm network started on bzz address: %v", self.hive.GetAddr()))
 
 	if self.pssEnabled {
 		pssparams := network.NewPssParams()
 		self.pss = network.NewPss(self.hive.Overlay, pssparams)
-		glog.V(logger.Info).Infof("Pss started: %v", self.pss)
+
```
Comment (zelig) [inline swarm/swarm.go:193]: I know you mentioned you put this in swarm, but this must be temporary. ok I see you left exactly this comment. All cool

File: swarm/swarm.go [hunk 6]
```diff
@@ -33,7 +33,7 @@ import (
 	"github.com/ethereum/go-ethereum/node"
 	"github.com/ethereum/go-ethereum/p2p"
 	"github.com/ethereum/go-ethereum/p2p/adapters"
-	"github.com/ethereum/go-ethereum/p2p/discover"
+	//"github.com/ethereum/go-ethereum/p2p/discover"
```
Comment (zelig) [inline swarm/swarm.go]: lots of commented out code in this file, lets remove them

**Review by zelig — CHANGES_REQUESTED** [review]
my favourite PR of recent times :)
makes me happy such great traction

File: swarm/network/pss.go (hunk 2 above)
Comment (nolash) [reply swarm/network/pss.go:478]: It's used by swarm/swarm.go, convenience test for implementers, will be removed when we're ready for productions. It's ok no?

File: swarm/network/pss.go (hunk 1 above)
Comment (nolash) [reply swarm/network/pss.go]: It might have been related to testing problems. I'll change it back, I can't remember if there was a better reason.
